### ISSUE TITLE
Regarding implementation on Franka

### ISSUE THREAD

**Issue by ehaller00** [openvla/openvla#265]
Hi OpenVLA Team,

Thank you so much for your work on OpenVLA, it’s really exciting, and I’m very interested in trying it out both in simulation and on a Franka Research 3 robot. I would like to ask some question to some issues I encountered so far.

- When using the model with a Franka arm, do I need to apply any action transformation steps, or can I use the calculated action (except for the gripper) directly with the robot? Are the steps basically the same as in the example for the WindowX robot you provided? 

- Can the model produce meaningful actions in a zero-shot setting with the Franka Arm, or is fine-tuning required for the model to behave meaningful in a new environment? What can I realistically expect from a zero-shot approach? At the moment, without fine-tuning the robot in LIBERO behaves a bit random with my settings. However, I saw that with some action processing steps, openvla-7b achieved around a 5–6% success rate on certain tasks using the GoogleBot in SimplerEnv.

Thanks again in advance and I really appreciate your time and support!

Best regards,

**Comment by ghost** [comment]
From my understanding, zero-shot performance is not good unless your setup closely matches one from the training data. But with a bit of fine-tuning (around 100 episodes ) you should get decent results.

The model predicts a 7D delta action: position (Δx, Δy, Δz), orientation (Δroll, Δpitch, Δyaw), and a binary gripper command. These are relative, not absolute.

To use them, apply the deltas to the current end-effector pose. Than you can use inverse kinematics to move the robot. Actions are normalized during training using per-dataset stats, so you’ll need to unnormalize them before execution using the correct unnorm_key. If you perform fine-tuneing, use your dataset’s stats.

Also note: coordinate frames differ by dataset. For example, in BridgeData V2:
ΔX = forward/back, ΔY = left/right, ΔZ = up/down (in the end-effector frame).
So, proper frame mapping is important.

Hope that helps!
Martin

**Comment by ehaller00** [comment]
Hi Martin, yes this helps a lot, thanks!

Maybe some follow-up questions for controlling the Franka FR3:

- Should I use `MoveIt` to do the inverse kinematics for me through its API using the `moveit_py` package, or is it recommended to do the inverse kinematics manually to directly control the joint positions?  
- Is there another approach that would let me control the physical robot efficiently with OpenVLA?

We are trying to get some Python script with `ROS 2 - Humble` controlling our `FR3`, but we have encountered some issues with the robot arm performing very inefficient trajectories to get to the specified position using the following script:

```python
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient

from moveit_msgs.action import MoveGroup
from moveit_msgs.msg import MotionPlanRequest, Constraints, PositionConstraint, OrientationConstraint, RobotState
from geometry_msgs.msg import PoseStamped
from shape_msgs.msg import SolidPrimitive
from builtin_interfaces.msg import Duration
import tf2_ros
import tf_transformations
import copy


class DeltaPoseMover(Node):
    def __init__(self):
        super().__init__('delta_pose_mover')

        self.group_name = "fr3_arm"
        self.planner_id = ""  # Leave empty for default
        self.ee_frame = "fr3_link7"
        self.base_frame = "fr3_link0"

        self.move_group_action = ActionClient(self, MoveGroup, '/move_action')
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)

        self.timer = self.create_timer(2.0, self.send_motion_goal)

    def send_motion_goal(self):
        self.timer.cancel()

        self.get_logger().info("Waiting for MoveGroup action server...")
        self.move_group_action.wait_for_server()

        try:
            transform = self.tf_buffer.lookup_transform(
                self.base_frame,
                self.ee_frame,
                rclpy.time.Time()
            )
        except Exception as e:
            self.get_logger().error(f"TF error: {e}")
            return

        # Get current pose and apply deltas
        current_pose = transform.transform
        new_pose = PoseStamped()
        new_pose.header.frame_id = self.base_frame
        new_pose.pose.position.x = current_pose.translation.x  + 0.2 # Apply translation delta
        new_pose.pose.position.y = current_pose.translation.y
        new_pose.pose.position.z = current_pose.translation.z

        quat = [
            current_pose.rotation.x,
            current_pose.rotation.y,
            current_pose.rotation.z,
            current_pose.rotation.w,
        ]
        roll, pitch, yaw = tf_transformations.euler_from_quaternion(quat)
        roll += 0.2  # Apply rotation delta

        new_quat = tf_transformations.quaternion_from_euler(roll, pitch, yaw)
        new_pose.pose.orientation.x = new_quat[0]
        new_pose.pose.orientation.y = new_quat[1]
        new_pose.pose.orientation.z = new_quat[2]
        new_pose.pose.orientation.w = new_quat[3]

        # Construct motion goal
        request = MotionPlanRequest()
        request.group_name = self.group_name
        request.goal_constraints.append(Constraints())
        request.goal_constraints[0].position_constraints.append(PositionConstraint())
        request.goal_constraints[0].orientation_constraints.append(OrientationConstraint())

        pos_constraint = request.goal_constraints[0].position_constraints[0]
        pos_constraint.header.frame_id = self.base_frame
        pos_constraint.link_name = self.ee_frame
        pos_constraint.target_point_offset.x = 0.0
        pos_constraint.target_point_offset.y = 0.0
        pos_constraint.target_point_offset.z = 0.0

        primitive = SolidPrimitive()
        primitive.type = SolidPrimitive.BOX
        primitive.dimensions = [0.01, 0.01, 0.01]  # Tight tolerance
        pos_constraint.constraint_region.primitives.append(primitive)
        pos_constraint.constraint_region.primitive_poses.append(new_pose.pose)
        pos_constraint.weight = 1.0

        orient_constraint = request.goal_constraints[0].orientation_constraints[0]
        orient_constraint.header.frame_id = self.base_frame
        orient_constraint.link_name = self.ee_frame
        orient_constraint.orientation = new_pose.pose.orientation
        orient_constraint.absolute_x_axis_tolerance = 0.01
        orient_constraint.absolute_y_axis_tolerance = 0.01
        orient_constraint.absolute_z_axis_tolerance = 0.01
        orient_constraint.weight = 1.0

        goal = MoveGroup.Goal()
        goal.request = request
        goal.planning_options.plan_only = False
        goal.planning_options.look_around = False
        goal.planning_options.replan = True

        self.get_logger().info("Sending motion plan goal...")
        self.move_group_action.send_goal_async(goal, feedback_callback=self.feedback_cb).add_done_callback(self.result_cb)

    def feedback_cb(self, feedback_msg):
        self.get_logger().info("Planning feedback received.")

    def result_cb(self, future):
        goal_handle = future.result()

        if not goal_handle.accepted:
            self.get_logger().error('Goal was rejected by MoveGroup.')
            return

        self.get_logger().info('Goal accepted. Waiting for result...')

        get_result_future = goal_handle.get_result_async()
        get_result_future.add_done_callback(self.result_response_cb)

    def result_response_cb(self, future):
        result = future.result().result

        if result.error_code.val == result.error_code.SUCCESS:
            self.get_logger().info("Motion succeeded!")
        else:
            self.get_logger().error(f"Motion failed: error code {result.error_code.val}")




def main(args=None):
    rclpy.init(args=args)
    node = DeltaPoseMover()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

```

Online, we found some examples using `moveit_py` we wanted to try it . Because the binary installation of `MoveIt2` for `ROS Humble` doesn't have the `moveit_py` package, we are now trying to install a newer version, `MoveIt 2 - 2.7.4 (Iron branch)`, from source. However, it is not working properly, and we are in the process of manually fixing issues such as the joint_limits.yaml can't be found or missing acceleration limits, etc.

**Comment by ghost** [comment]
I'm currently using C++ along with the Franka Control Interface (FCI) libraries, running everything without ROS. However, I believe this approach should also work with MoveIt.

The core idea I'm pursuing is to read the robot’s current state (x,y,z, r,p,y) add the actions to all 6 dimensions to reach a desired target state. Than I use a controller to generate a smooth and feasible trajectory from one state to the next. 

That said, I’m still in the early stages of implementation, so I’m exploring and learning as I go. :)

**Comment by lxsy-xcy** [comment]
I'm also experiencing this question. What un-normkey do you use to test the zero-shot performance on Franka?
